src/main/scala/com/fulcrumgenomics/vcf/ByIntervalListVariantContextIterator.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.vcf
+
+import java.util.NoSuchElementException
+
+import htsjdk.samtools.util.{CloseableIterator, Interval, IntervalList}
+import htsjdk.variant.variantcontext.VariantContext
+import htsjdk.variant.vcf.VCFFileReader
+import htsjdk.samtools.util.OverlapDetector
+
+
+
+object ByIntervalListVariantContextIterator {
+  /**
+    * Creates an iterator over variant contexts that overlap any interval in an interval list.
+    *
+    * If `useIndex` is true, this will query the VCF when moving to the next variant context, and so may be quite slow
+    * if we jump around the VCF a lot.  Otherwise, all variants will be read in and filtered appropriately.
+    */
+  def apply(reader: VCFFileReader,
+            intervalList: IntervalList,
+            useIndex: Boolean = true): Iterator[VariantContext] = {
+    if (useIndex) {
+      new IndexQuery(reader, intervalList)
+    }
+    else {
+      import scala.collection.JavaConversions.asScalaIterator
+      val detector = {
+        import scala.collection.JavaConverters._
+        val intervals = intervalList.uniqued().iterator().toList.asJava
+        val d = new OverlapDetector[Interval](0, 0)
+        d.addAll(intervals, intervals)
+        d
+      }
+      reader.iterator().toIterator.filter { ctx => detector.overlapsAny(ctx) }
+    }
+  }
+}
+
+private class IndexQuery(reader: VCFFileReader,
+                         intervalList: IntervalList)
+  extends Iterator[VariantContext] {
+  private var iterator: Option[CloseableIterator[VariantContext]] = None
+  private val intervals = intervalList.uniqued().iterator()
+
+  def hasNext(): Boolean = {
```

Comment:
I know this is ported over (and probably from something I wrote!!), but can we do the following refactor please:
1) In the constructor call advance() once to prime the iterator
2) remove the advance() call from hasNext()
3) change the next() method to:
```scala
def next(): VariantContext = {
  this.iterator match {
    case None => throw new NoSuchElementException("Called next when hasNext is false")
    case Some(iter) => yieldAndThen(iter.next()) { advance() }
  }
}
```